common/src/main/java/net/william278/husksync/database/MongoDbDatabase.java

```diff
@@ -0,0 +1,357 @@
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to establish a connection to the MongoDB database. " +
+                    "Please check the supplied database credentials in the config file", e);
+        }
+    }
+
+    /**
+     * Ensure a {@link User} has an entry in the database and that their username is up-to-date
+     *
+     * @param user The {@link User} to ensure
+     */
+    @Blocking
+    @Override
+    public void ensureUser(@NotNull User user) {
+        getUser(user.getUuid()).ifPresentOrElse(
+                existingUser -> {
+                    if (!existingUser.getUsername().equals(user.getUsername())) {
+                        // Update a user's name if it has changed in the database
+                        try {
+                            Document filter = new Document("uuid", existingUser.getUuid().toString());
+                            Document doc = mongoCollectionHelper.getCollection(usersTable).find(filter).first();
+
+                            Bson updates = Updates.set("uuid", user.getUuid().toString());
+                            mongoCollectionHelper.updateDocument(usersTable, doc, updates);
+                        } catch (MongoException e) {
+                            plugin.log(Level.SEVERE, "Failed to insert a user into the database", e);
+                        }
+                    }
+                },
+                () -> {
+                    // Insert new player data into the database
+                    try {
+                        Document doc = new Document("uuid", user.getUuid().toString()).append("username", user.getUsername());
+                        mongoCollectionHelper.insertDocument(usersTable, doc);
+                    } catch (MongoException e) {
+                        plugin.log(Level.SEVERE, "Failed to insert a user into the database", e);
+                    }
+                }
+        );
+    }
+
+    /**
+     * Get a player by their Minecraft account {@link UUID}
+     *
+     * @param uuid Minecraft account {@link UUID} of the {@link User} to get
+     * @return An optional with the {@link User} present if they exist
+     */
+    @Blocking
+    @Override
+    public Optional<User> getUser(@NotNull UUID uuid) {
+        Document filter = new Document("uuid", uuid);
+        Document doc = mongoCollectionHelper.getCollection(usersTable).find(filter).first();
+        if (doc != null) {
+            return Optional.of(new User(UUID.fromString(doc.getString("uuid")),
+                    doc.getString("username")));
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get a user by their username (<i>case-insensitive</i>)
+     *
+     * @param username Username of the {@link User} to get (<i>case-insensitive</i>)
+     * @return An optional with the {@link User} present if they exist
+     */
+    @Blocking
+    @Override
+    public Optional<User> getUserByName(@NotNull String username) {
```

Comment:
add NotNull annotation

Reply:
where do you want the notnull??? it already exists